pkg/controller/job/job_controller.go

```diff
@@ -1322,7 +1352,10 @@ func (jm *Controller) recordJobFinished(job *batch.Job, finishedCond *batch.JobC
 			jm.recorder.Event(job, v1.EventTypeWarning, "TooManySucceededPods", "Too many succeeded pods running after completion count reached")
 		}
 		jm.recorder.Event(job, v1.EventTypeNormal, "Completed", "Job completed")
+		// The succeeded metric with an empty reason is for the backward compatibility.
 		metrics.JobFinishedNum.WithLabelValues(completionMode, "succeeded", "").Inc()
+		succeededReason := jobFinishedNumReasonLabel(job)
+		metrics.JobFinishedNum.WithLabelValues(completionMode, "succeeded", succeededReason).Inc()
```

Comment:
> I guess this we should only bump if the feature gate is enabled.

That makes sense. This is my bad.

> However, bumping the same metric twice for a different set of labels seems tricky, not sure if this violates: the Stable metrics expectations: "Furthermore, Stable metrics are guaranteed to not change in respect to labels" from: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3498-extending-stability#stable-metrics. 

I agree with you, but we can take the only way with backward compatibility since the empty reason is defined as a job that succeeded regardless of the reasons: https://github.com/kubernetes/kubernetes/blob/cc5362ebc17e1376fa79b510f7f354dbffe7f92e/pkg/controller/job/metrics/metrics.go#L58-L59

> Has this been discussed in KEP, with sig-instrumentation folks or with API reviewers?

Yes, we have mentioned these new reasons, and those reasons were accepted, but I'm not sure if reviewers agree with bumping the same metrics twice.

https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/3998-job-success-completion-policy/README.md#monitoring-requirements

Or, maybe we should add a new metric separate from `jobs_finished_total` like `jobs_finished_total_v2`. @mimowo @kannon92 WDYT?

Anyway, I agree that we should have a consensus with API reviewers.

Reply:
> @mimowo @kannon92 WDYT?

My preference is not to bump it twice. I think in some tools like Prometheus you may collapse all label values, then it would show too large values.

These are the options in the order of my preference:
1. bump it once for the `reason` field indicating this is due to success policy (breaking compatibility). Maybe breaking compatibility on the set of label values is acceptable, probably depends on the interpretation of "Stable metrics are guaranteed to not change in respect to labels".
2. `jobs_succeeded_total`, with label `reason` indicating `completions` or `successPolicy` 
3. `jobs_finished_total_v2` 

WDYT @deads2k @alculquicondor ?

Reply:
It sounds more attractive to the second option, "jobs_succeeded_total".

Reply:
I was told that metrics should go in beta actually. But I think having them in alpha would be useful so not sure if that is actually true.

Reply:
Yeah, I think metrics can go in Beta, so we may not update them in this release cycle at all, and return to the discussion for the next release cycle.

Reply:
I see. I'm ok with removing the new metric in the alpha stage.
But let me wait for the response from API reviewer @deads2k.